packages/sdk/src/EventSenderEngine.ts

```diff
@@ -0,0 +1,116 @@
+import { Value } from './Value';
+import { FetchBuilder, TimeUnit } from '@spotify-confidence/client-http';
+interface Event {
+  eventDefinition: string;
+  eventTime: string;
+  payload: Value.Struct;
+}
+
+interface EventBatch {
+  sendTime: string;
+  clientSecret: string;
+  events: Event[];
+}
+interface PublishError {
+  index: number;
+  reason: string;
+  message: string;
+}
+
+export interface EventSenderEngineOptions {
+  clientSecret: string;
+  maxBatchSize: number;
+  flushTimeoutMilliseconds: number;
+  rateLimitRps?: number;
+  fetchImplementation: FetchImplementation;
+  region: 'eu' | 'us';
+}
+
+type FetchImplementation = typeof fetch;
+
+export class EventSenderEngine {
+  private readonly writeQueue: Event[] = [];
+  private readonly flushTimeoutMilliseconds: number;
+  private readonly clientSecret: string;
+  private readonly maxBatchSize: number;
+  private readonly fetchImplementation: FetchImplementation;
+  private readonly publishUrl: string;
+  private pendingFlush: undefined | ReturnType<typeof setTimeout>;
+
+  constructor({
+    clientSecret,
+    maxBatchSize,
+    flushTimeoutMilliseconds,
+    fetchImplementation,
+    region,
+    rateLimitRps,
+  }: EventSenderEngineOptions) {
+    this.publishUrl = `https://events.${region}.confidence.dev/v1/events:publish`;
+    this.clientSecret = clientSecret;
+    this.maxBatchSize = maxBatchSize;
+    this.flushTimeoutMilliseconds = flushTimeoutMilliseconds;
+    const fetchBuilder = new FetchBuilder()
+      .limitPending(1000)
+      .rejectNotOk()
+      .timeout(30 * TimeUnit.MINUTE)
+      .retry({ delay: 5 * TimeUnit.SECOND, backoff: 2, maxDelay: 5 * TimeUnit.MINUTE, jitter: 0.2 })
+      .rejectOn(({ status }) => status >= 500 || status === 429);
+
+    if (rateLimitRps) {
+      fetchBuilder.rateLimit(rateLimitRps);
+    }
+    this.fetchImplementation = fetchBuilder
+      // update send-time before sending
+      .modifyRequest(async request => {
+        if (request.method === 'POST') {
+          const body = JSON.stringify({ ...(await request.json()), sendTime: new Date().toISOString() });
+          return new Request(request, { body });
+        }
+        return request;
+      })
+      .build(fetchImplementation);
+  }
+
+  send(context: Value.Struct, name: string, message?: Value.Struct): void {
+    this.writeQueue.push({
+      eventDefinition: `eventDefinitions/${name}`,
+      eventTime: new Date().toISOString(),
```

Comment:
nit: Worth sending a clock interface to be able to unit test the time things?

Reply:
We can use fake timers to do that